src/AspNet.Security.OAuth.Zoho/ZohoAuthenticationHandler.cs

```diff
@@ -0,0 +1,166 @@
+        requestMessage.Version = Backchannel.DefaultRequestVersion;
+
+        using var response = await Backchannel.SendAsync(requestMessage, HttpCompletionOption.ResponseHeadersRead, Context.RequestAborted);
+        if (!response.IsSuccessStatusCode)
+        {
+            await Log.UserProfileErrorAsync(Logger, response, Context.RequestAborted);
+            throw new HttpRequestException("An error occurred while retrieving the user profile.");
+        }
+
+        using var payload = JsonDocument.Parse(await response.Content.ReadAsStringAsync(Context.RequestAborted));
+
+        var principal = new ClaimsPrincipal(identity);
+        var context = new OAuthCreatingTicketContext(principal, properties, Context, Scheme, Options, Backchannel, tokens, payload.RootElement);
+        context.RunClaimActions();
+
+        await Events.CreatingTicket(context);
+        return new AuthenticationTicket(context.Principal!, context.Properties, Scheme.Name);
+    }
+
+    protected override async Task<OAuthTokenResponse> ExchangeCodeAsync(OAuthCodeExchangeContext context)
+    {
+        var nameValueCollection = new Dictionary<string, string?>
+        {
+            ["client_id"] = Options.ClientId,
+            ["client_secret"] = Options.ClientSecret,
+            ["code"] = context.Code,
+            ["redirect_uri"] = context.RedirectUri,
+            ["grant_type"] = "authorization_code"
+        };
+
+        if (context.Properties.Items.TryGetValue(OAuthConstants.CodeVerifierKey, out var codeVerifier))
+        {
+            nameValueCollection.Add(OAuthConstants.CodeVerifierKey, codeVerifier!);
+            context.Properties.Items.Remove(OAuthConstants.CodeVerifierKey);
+        }
+
+        var tokenEndpoint = CreateEndpoint(ZohoAuthenticationDefaults.TokenPath);
+        using var request = new HttpRequestMessage(HttpMethod.Post, tokenEndpoint);
+        request.Headers.Accept.Add(new MediaTypeWithQualityHeaderValue(MediaTypeNames.Application.Json));
+        request.Content = new FormUrlEncodedContent(nameValueCollection);
+        request.Version = Backchannel.DefaultRequestVersion;
+
+        using var response = await Backchannel.SendAsync(request, Context.RequestAborted);
+        if (!response.IsSuccessStatusCode)
+        {
+            await Log.ExchangeCodeErrorAsync(Logger, response, Context.RequestAborted);
+            return OAuthTokenResponse.Failed(new Exception("An error occurred while retrieving an access token."));
+        }
+
+        var payload = JsonDocument.Parse(await response.Content.ReadAsStringAsync(Context.RequestAborted));
+
+        return OAuthTokenResponse.Success(payload);
+    }
+
+    private string CreateEndpoint(string path)
+    {
+        var location = Context.Request.Query["location"];
+
+        var domain = GetDomain(location.ToString());
+
+        return $"{domain}{path}";
+    }
+
+    private static string GetDomain(string location)
```

Comment:
Looks like `CreateEndpoint` is the only method that calls it. Maybe we can simplify the code to merge the two helpers?

Please also add a code comment indicating that we deliberately don't use the `accounts-server` parameter and that we fall back to the US region when the region isn't recognized for security reasons (blindly using `accounts-server` without checking it would allow stealing the client secret very easily). Just in case future us would forget this little detail 🤣